File: src/main/java/com/kralofsky/citybikes/bot/CitybikeTelegramBot.java
package com.kralofsky.citybikes.bot;

import com.google.common.collect.ImmutableMap;
import com.kralofsky.citybikes.bot.util.ExternalAbility;
import com.kralofsky.citybikes.config.Values;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.telegram.abilitybots.api.bot.AbilityBot;
import org.telegram.abilitybots.api.bot.BaseAbilityBot;
import org.telegram.abilitybots.api.db.DBContext;
import org.telegram.abilitybots.api.objects.Ability;
import org.telegram.abilitybots.api.objects.Reply;

import java.lang.reflect.Field;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import java.util.stream.Stream;


@Component
@Slf4j
public class CitybikeTelegramBot extends AbilityBot {
    static final String DEFAULT_COMMAND = BaseAbilityBot.DEFAULT;

    private Values values;

    @Autowired
    public CitybikeTelegramBot(Values values, DBContext dbContext, Collection<? extends ExternalAbility> externalAbilities) {
        super(values.getBotToken(), values.getBotUsername(), dbContext);
        this.values = values;
        log.info("Initialize Bot " + values.getBotUsername());

        externalAbilities.forEach(a -> a.init(this));
        Collection<Ability> newAbilities = externalAbilities.stream().map(ExternalAbility::getAbility).collect(Collectors.toList());

        try {
            registerAbilities(newAbilities);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            log.error("Error while registering Abilities", e);
            throw new RuntimeException(e);
        }

    }

    @Override
    public int creatorId() {
        return values.getCreatorId();
    }

    // TODO This is a hack that uses reflection to write to private fields of the superclass...
    private void registerAbilities(Collection<Ability> newAbilities) throws NoSuchFieldException, IllegalAccessException {
        Class bab = this.getClass().getSuperclass().getSuperclass();
        Field abilitiesField = bab.getDeclaredField("abilities");
        abilitiesField.setAccessible(true);
        Map<String, Ability> existingAbilities = (Map<String, Ability>) abilitiesField.get(this);

        Map<String, Ability> abilities = Stream
                .concat(existingAbilities.values().stream(), newAbilities.stream())
                .collect(ImmutableMap::<String, Ability>builder,
                        (b, a) -> b.put(a.name(), a),
                        (b1, b2) -> b1.putAll(b2.build()))
                .build();

        abilitiesField.set(this, abilities);

        Stream<Reply> newReplies = abilities.values().stream().flatMap(ability -> ability.replies().stream());

        Field repliesField = bab.getDeclaredField("replies");
        repliesField.setAccessible(true);
        List<Reply> existingReplies = (List<Reply>) repliesField.get(this);

        List<Reply> replies = Stream.concat(existingReplies.stream(), newReplies).collect(Collectors.toList());

        repliesField.set(this, replies);
    }
}

File: src/main/java/com/kralofsky/citybikes/bot/CitybikeTelegramBot.kt
package com.kralofsky.citybikes.bot

import com.google.common.collect.ImmutableMap
import com.kralofsky.citybikes.bot.util.ExternalAbility
import com.kralofsky.citybikes.config.Values
import com.kralofsky.citybikes.config.logger
import org.springframework.stereotype.Component
import org.telegram.abilitybots.api.bot.AbilityBot
import org.telegram.abilitybots.api.bot.BaseAbilityBot
import org.telegram.abilitybots.api.db.DBContext
import org.telegram.abilitybots.api.objects.Ability
import org.telegram.abilitybots.api.objects.Reply
import java.util.stream.Collectors
import java.util.stream.Stream


@Component
class CitybikeTelegramBot (
        private val values: Values,
        dbContext: DBContext,
        externalAbilities: Collection<ExternalAbility>
) : AbilityBot(values.botToken, values.botUsername, dbContext) {
    private final val log = logger()

    init {
        log.info("Initialize Bot " + values.botUsername)

        externalAbilities.forEach { a -> a.init(this) }

        registerAbilities(externalAbilities.map { it.ability })
    }

    override fun creatorId(): Int {
        return values.creatorId
    }

    // TODO This is a hack that uses reflection to write to private fields of the superclass...
    private fun registerAbilities(newAbilities: Collection<Ability>) {
        val bab = this.javaClass.superclass.superclass
        val abilitiesField = bab.getDeclaredField("abilities")
        abilitiesField.isAccessible = true
        val existingAbilities = abilitiesField.get(this) as Map<String, Ability>

        val abilities = (existingAbilities.values + newAbilities).map { it.name() to it }.toMap()

        abilitiesField.set(this, abilities)

        val newReplies = abilities.values.flatMap { it.replies() }

        val repliesField = bab.getDeclaredField("replies")
        repliesField.isAccessible = true
        val existingReplies = repliesField.get(this) as List<Reply>

        val replies = (existingReplies + newReplies)

        repliesField.set(this, replies)
    }

    companion object {
        val DEFAULT_COMMAND = BaseAbilityBot.DEFAULT
    }
}
